source/extensions/common/http_wasm/context.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+
+#include <atomic>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <string_view>
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/buffer/buffer.h"
+#include "envoy/http/filter.h"
+#include "envoy/stats/sink.h"
+#include "source/common/buffer/buffer_impl.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+
+#include "source/extensions/common/http_wasm/guest.h"
+#include "source/extensions/common/http_wasm/vm_runtime.h"
+#include "source/extensions/common/http_wasm/http_wasm_common.h"
+#include "source/extensions/common/http_wasm/guest_config.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HttpWasm {
+
+class GuestConfigHandle;
+class GuestConfig;
+class Guest;
+class GuestHandle;
+
+using GuestConfigSharedPtr = std::shared_ptr<GuestConfig>;
+using GuestConfigHandleSharedPtr = std::shared_ptr<GuestConfigHandle>;
+using GuestHandleSharedPtr = std::shared_ptr<GuestHandle>;
+
+class WasmBuffer {
+public:
+  WasmBuffer() = default;
+
+  size_t size() const;
+  int64_t copyTo(void* ptr, uint64_t size);
+  WasmResult copyFrom(std::string_view data, size_t length);
+
+  void clear() {
+    const_buffer_instance_ = nullptr;
+    buffer_instance_ = nullptr;
+  }
+
+  WasmBuffer* set(::Envoy::Buffer::Instance* buffer_instance) {
+    clear();
+    buffer_instance_ = buffer_instance;
+    const_buffer_instance_ = buffer_instance;
+    return this;
+  }
+  WasmBuffer* set(Buffer::OwnedImpl* buffer_instance) {
+    clear();
+    buffer_instance_ = buffer_instance;
+    const_buffer_instance_ = buffer_instance;
+    return this;
+  }
+  WasmBuffer* set(const ::Envoy::Buffer::Instance* buffer_instance) {
+    clear();
+    const_buffer_instance_ = buffer_instance;
+    return this;
+  }
+
+private:
+  const ::Envoy::Buffer::Instance* const_buffer_instance_{};
+  ::Envoy::Buffer::Instance* buffer_instance_{};
+  uint64_t bytes_to_skip_ = 0;
+};
+
+// A context which will be the target of callbacks for a particular session
+// e.g. a handler of a stream.
+class Context : public Logger::Loggable<Logger::Id::wasm>,
+                public Http::StreamFilter,
+                public std::enable_shared_from_this<Context> {
+public:
+  Context() = default; // Testing.
+  Context(Guest* guest, GuestConfigSharedPtr& initialized_guest);
+
+  ~Context() override;
+
+  Guest* guest() const { return guest_; }
+  uint32_t id() const { return id_; }
+  Context* parent_context() const { return parent_context_; }
```

Comment:
do we need "parent context" anymore? It is confusing as we don't have "parent context" in http-wasm level, and I think we can simply make each stream context has a reference to guest